baybe/searchspace/discrete.py

```diff
@@ -298,52 +308,113 @@ def from_simplex(
 
-        # Get the minimum sum contributions to come in the upcoming joins (the first
-        # item is the minimum possible sum of all parameters starting from the
+        # Get the minimum sum contributions to come in the upcoming joins (the
+        # first item is the minimum possible sum of all parameters starting from the
         # second parameter, the second item is the minimum possible sum starting from
         # the third parameter, and so on ...)
-        min_upcoming = np.cumsum(min_values[:0:-1])[::-1]
+        min_sum_upcoming = np.cumsum(min_values[:0:-1])[::-1]
+
+        # Get the min/max number of nonzero values to come in the upcoming joins (the
+        # first item is the min/max number of nonzero parameters starting from the
+        # second parameter, the second item is the min/max number starting from
+        # the third parameter, and so on ...)
+        min_nonzero_upcoming = np.cumsum((np.asarray(min_values) > 0.0)[:0:-1])[::-1]
+        max_nonzero_upcoming = np.cumsum((np.asarray(max_values) > 0.0)[:0:-1])[::-1]
 
         # Incrementally build up the space, dropping invalid configuration along the
-        # way. More specifically: after having cross-joined a new parameter, there must
-        # be enough "room" left for the remaining parameters to fit. Hence,
-        # configurations of the current parameter subset that exceed the desired
-        # total value minus the minimum contribution to come from the yet to be added
-        # parameters can be already discarded.
-        for i, (param, min_to_go) in enumerate(
-            zip_longest(simplex_parameters, min_upcoming, fillvalue=0)
+        # way. More specifically:
+        # * After having cross-joined a new parameter, there must
+        #   be enough "room" left for the remaining parameters to fit. That is,
+        #   configurations of the current parameter subset that exceed the desired
+        #   total value minus the minimum contribution to come from the yet-to-be-added
+        #   parameters can be already discarded, because it is already clear that
+        #   the total sum will be exceeded once all joins are completed.
+        # * Analogously, there must be enough "nonzero slots" left for the yet to be
+        #   joined parameters, i.e. parameter subset configurations can be discarded
+        #   where the number of nonzero parameters already exceeds the maximum number
+        #   of nonzeros minus the number of nonzeros to come, because it is already
+        #   clear that the maximum will be exceeded once all joins are completed.
+        # * Similarly, it can be verified for each parameter that there are still
+        #   enough nonzero parameters to come to even reach the minimum
+        #   desired number of nonzero after all joins.
+        for i, (
+            param,
+            min_sum_to_go,
+            min_nonzero_to_go,
+            max_nonzero_to_go,
+        ) in enumerate(
+            zip_longest(
```

Comment:
Why is the `zip_longest` necessary, shouldn't all of these have the same length?

Reply:
Or this due to the fact that there might be other parameters like Categorical ones that do not contribute here?

Reply:
hthink it has to do with cumsum yielding a length-1 of the pbject you pass to it, that is prob trvially 0 hence the fillup